### PR TITLE
fix(SubscriberHistory): add campaign preview in grid

### PR DESCRIPTION
**Pull request by fcoronelmakingsense**
- show preview when has preview image if not only show the link to the campaign report.
- add test to verify the mapping when has preview

image

### CODE REVIEW

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js [hunk 1]
```diff
@@ -55,9 +55,15 @@ const SubscriberHistory = ({ location, dependencies: { dopplerApiClient } }) =>
         if (!response.success) {
           setState({ loading: false });
         } else {
+          const sentCampaigns = response.value.items.map((campaign) => {
+            return {
+              ...campaign,
+              imgPreview: campaign.links.find((x) => x.rel === '/docs/rels/get-campaign-preview'),
```
Comment (andresmoschini) [inline src/components/Reports/SubscriberHistory/SubscriberHistory.js]: I am not sure if it is so simple, for example, a link could have multiple relations, for example: `"rel": "/docs/rels/get-campaign-preview preview image"`

What do you think about adding a helper for this?

```suggestion
              imgPreview: searchLinkByRel(campaign, '/docs/rels/get-campaign-preview'),
```

Comment (andresmoschini) [reply src/components/Reports/SubscriberHistory/SubscriberHistory.js]: What happens when there is no link with specified rel?

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js [hunk 2]
```diff
@@ -123,7 +129,27 @@ const SubscriberHistory = ({ location, dependencies: { dopplerApiClient } }) =>
...
+                                </a>
+                              </div>
+                            ) : (
+                              <a
+                                href={`https://reports2.fromdoppler.com/Dashboard.aspx?idCampaign=${campaign.campaignId}`}
+                              >
+                                {campaign.campaignName}
+                              </a>
+                            )}
+                            {}
```
Comment (andresmoschini) [inline src/components/Reports/SubscriberHistory/SubscriberHistory.js]: What is this?

File: src/components/Reports/SubscriberHistory/SubscriberHistory.test.js [hunk 3]
```diff
@@ -236,4 +241,57 @@ describe('SubscriberHistory component', () => {
       expect(getByText('subscriber_history.delivery_status.notOpened')).toBeInTheDocument();
     });
   });
+
+  it('should load campaign preview img', async () => {
```
Comment (cbernat) [inline src/components/Reports/SubscriberHistory/SubscriberHistory.test.js]: Perhaps add another test for when tooltip is not found?

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js [hunk 4]
```diff
@@ -123,7 +129,27 @@ const SubscriberHistory = ({ location, dependencies: { dopplerApiClient } }) =>
...
-                          <td>{campaign.campaignName}</td>
+                          <td>
+                            {campaign.imgPreview ? (
+                              <div className="dp-tooltip-container">
+                                <a
+                                  href={`https://reports2.fromdoppler.com/Dashboard.aspx?idCampaign=${campaign.campaignId}`}
+                                >
+                                  {campaign.campaignName}
+                                  <div className="dp-tooltip-block">
+                                    <img src={campaign.imgPreview.href} alt="" />
```
Comment (cbernat) [inline src/components/Reports/SubscriberHistory/SubscriberHistory.js]:
```suggestion
                                    <img src={campaign.imgPreview.href} alt="Campaign Preview" />
```
Or better add this into a resourced text.

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js (hunk 1 above)
Comment (fcoronelmakingsense) [reply src/components/Reports/SubscriberHistory/SubscriberHistory.js]: we talk with @cbernat in the api the rel is this for the preview, if not dont have preview and only show the link to campaign report

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js (hunk 2 above)
Comment (fcoronelmakingsense) [reply src/components/Reports/SubscriberHistory/SubscriberHistory.js]: it's a mistake

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js (hunk 1 above)
Comment (cbernat) [reply src/components/Reports/SubscriberHistory/SubscriberHistory.js]: I think that @fcoronelmakingsense has this case covered, for not existing preview link.

Comment (fcoronelmakingsense) [reply src/components/Reports/SubscriberHistory/SubscriberHistory.js]: fixed now!

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js (hunk 2 above)
Comment (fcoronelmakingsense) [reply src/components/Reports/SubscriberHistory/SubscriberHistory.js]: fixed

File: src/components/Reports/SubscriberHistory/SubscriberHistory.test.js (hunk 3 above)
Comment (fcoronelmakingsense) [reply src/components/Reports/SubscriberHistory/SubscriberHistory.test.js]: ready

File: src/components/Reports/SubscriberHistory/SubscriberHistory.test.js [hunk 5]
```diff
@@ -294,4 +294,50 @@ describe('SubscriberHistory component', () => {
       expect(container.querySelector('.dp-tooltip-block')).toBeInTheDocument();
     });
   });
+
+  it('should load campaign without preview img', async () => {
```
Comment (cbernat) [inline src/components/Reports/SubscriberHistory/SubscriberHistory.test.js]: ❤️

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js (hunk 4 above)
Comment (fcoronelmakingsense) [reply src/components/Reports/SubscriberHistory/SubscriberHistory.js]: fixed

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js [hunk 6]
```diff
@@ -28,6 +28,10 @@ const getDeliveryStatusCssClassName = (deliveryStatus) => {
   return deliveryCssClass;
 };
 
+const searchLinkByRel = (items, rel) => {
+  return items.find((x) => x.rel.includes(rel));
```
Comment (andresmoschini) [inline src/components/Reports/SubscriberHistory/SubscriberHistory.js]: Great, in the future I think that _searchLinkByRel_ could be improved to be more robust, for example, splitting the string using spaces, but by the moment it is fine and it is easy to modify.

File: src/components/Reports/SubscriberHistory/SubscriberHistory.js [hunk 7]
```diff
@@ -55,9 +59,15 @@ const SubscriberHistory = ({ location, dependencies: { dopplerApiClient } }) =>
         if (!response.success) {
           setState({ loading: false });
         } else {
+          const sentCampaigns = response.value.items.map((campaign) => {
+            return {
+              ...campaign,
+              imgPreview: searchLinkByRel(campaign.links, '/docs/rels/get-campaign-preview'),
+            };
```
Comment (andresmoschini) [inline src/components/Reports/SubscriberHistory/SubscriberHistory.js]: In my opinion, this mapping work is a responsibility of _dopplerApiClient_.

Comment (fcoronelmakingsense) [reply src/components/Reports/SubscriberHistory/SubscriberHistory.js]: i moved now to the service!

File: src/services/doppler-api-client.double.ts [hunk 8]
```diff
@@ -69,13 +69,11 @@ const campaignDeliveryItems = [...Array(100)].map((_, index) => {
...
-      {
-        href: 'http://dopplerfilesint.fromdoppler.net/Users/50018/Campaigns/33850437/33850437.png',
-        description: 'get preview',
-        rel: '/docs/rels/get-campaign-preview',
-      },
-    ],
+    imgPreview: {
+      href: 'http://dopplerfilesint.fromdoppler.net/Users/50018/Campaigns/33850437/33850437.png',
+      description: 'get preview',
+      rel: '/docs/rels/get-campaign-preview',
```
Comment (andresmoschini) [inline src/services/doppler-api-client.double.ts]: No convendría mapear solo el href?

Comment (fcoronelmakingsense) [reply src/services/doppler-api-client.double.ts]: si podria ser, pense el description podria llegar a servir para algo, pero si por ahi no

Comment (fcoronelmakingsense) [reply src/services/doppler-api-client.double.ts]: listo cambiado

**Pull request merged**